### ISSUE TITLE
sort() and  topk() behave weirdly when given large numbers

### ISSUE THREAD

**Issue by jasonleeinf** [pytorch/pytorch#3462]
```python
import torch
import numpy as np

batch_size = 5
dim = 10

### pytorch
a1 = torch.randn(batch_size, dim)
v1, r1 = a1.sort(dim=1)

inc = torch.ones(batch_size)[:,None]*99999999
a2 = inc + a1
v2, r2 = a2.sort(dim=1)

### numpy
a3 = a1.numpy()
r3 = np.argsort(a3, axis=1)

inc2 = np.ones(batch_size)[:,None]*99999999
a4 = inc2 + a3
r4 = np.argsort(a4, axis=1)
```

In the above, a matrix of large numbers, where every element is the same, is added to a matrix of small numbers (this shouldn't affect the ordering of the small matrix). Hence `r1, r2, r3, r4` should all be the same. However, `r2` is different from the rest, meaning adding a large numbers to the input makes `torch.sort()` behave weirdly. So does `torch.topk()`, presumably for the same reason.

I've been getting weird results with my beam search implementation in PyTorch, and this was the culprit. In beam search you have to sort a list of large negative numbers (log probabilities of candidate translation sentences). Can anyone suggest a quick workaround? Thanks!

**Comment by ezyang** [comment]
This is a floating point accuracy problem; at large numbers, float can't distinguish small differences. If your numbers are not too big, increasing precision to double can help solve your problem:

```
import torch

batch_size = 3
dim = 3

### pytorch
a1 = torch.DoubleTensor(batch_size, dim).normal_()
v1, r1 = a1.sort(dim=1)

inc = torch.DoubleTensor(batch_size).fill_(1)[:,None]*99999999
a2 = inc + a1
v2, r2 = a2.sort(dim=1)

print(r1)
print(r2)
```

Numpy does not have this problem because it defaults to double precision.
